python/source/state_representation/bind_cartesian_space.cpp

```diff
@@ -82,8 +86,15 @@ void cartesian_state(py::module_& m) {
   c.def("set_position", py::overload_cast<const Eigen::Vector3d&>(&CartesianState::set_position), "Setter of the position");
   c.def("set_position", py::overload_cast<const std::vector<double>&>(&CartesianState::set_position), "Setter of the position from a list");
   c.def("set_position", py::overload_cast<const double&, const double&, const double&>(&CartesianState::set_position), "Setter of the position from three scalar coordinates", "x"_a, "y"_a, "z"_a);
-  c.def("set_orientation", py::overload_cast<const Eigen::Vector4d&>(&CartesianState::set_orientation), "Setter of the orientation from a 4d vector of quaternion coeffiecients (w, x, y, z)");
-  c.def("set_orientation", py::overload_cast<const std::vector<double>&>(&CartesianState::set_orientation), "Setter of the orientation from a 4d list of quaternion coeffiecients (w, x, y, z)");
+  c.def("set_orientation", [](CartesianState &state, const py::object& quaternion) {
+    py::object PyQuaternion = py::module_::import("pyquaternion").attr("Quaternion");
+    if (!py::isinstance(quaternion, PyQuaternion)) {
+      throw std::invalid_argument("Input argument quaternion is not of type pyquaternion.quaternion.Quaternion");
+    }
+    state.set_orientation_coefficients(py::cast<Eigen::Vector4d>(quaternion.attr("elements")));
+  }, "Setter of the orientation attribute from a pyquaternion object");
+  c.def("set_orientation_coefficients", py::overload_cast<const Eigen::Vector4d&>(&CartesianState::set_orientation_coefficients), "Setter of the orientation from a 4d vector of quaternion coeffiecients (w, x, y, z)");
+  c.def("set_orientation_coefficients", py::overload_cast<const std::vector<double>&>(&CartesianState::set_orientation_coefficients), "Setter of the orientation from a 4d list of quaternion coeffiecients (w, x, y, z)");
```

Comment:
Would it be possible to use overloads to keep just the one function with different signatures? Or does the py::object capture every possible type? Something like:

```c++
py::overload_cast<const py::object& quaternion>([](CartesianState &state, const py::object& quaternion) {
   py::object PyQuaternion = ...
})
```

Otherwise, I suggest you bind a C++ class that is just a wrapper for PyQuaternion so that you have a proper type signature.

Reply:
First, I'm not sure that you can use lambda functions with `py::overload_cast` and second, I think everything would be captured by `py::object` indeed. I see another option if you prefer to keep just `set_orientation`:
- Have one function in python with three `isinstance` cases. This comes with the drawback of not having a nice automatically generated docstring as we have for the overloaded function

To be honest, I still like the way it is now, where we disambiguate between set_orientation and set_orientation_coefficients (as we do for the getter). Why do you not like that?

Reply:
Mainly I didn't like the two setters because it doesn't reflect the C++ usage. Setting with a quaternion object or with coefficients is done with the single signature `set_orientation` in C++. It will also break some components and other downstream usages, which I would like to avoid unless necessary.
The `is_instance` approach could work. Otherwise, the pyquaternion defines the `__getitem__(index)` method, which means it's addressable and iterable. So for any compatible py::object, you could just (try) to do `[coeff for coeff in obj]` to listify the elements.